File: src/main/java/oscar/cp/constraints/AllDiffFWC.java
/*******************************************************************************
 * This file is part of OscaR (Scala in OR).
 *   
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *  
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/gpl-3.0.html
 ******************************************************************************/
package oscar.cp.constraints;

import oscar.cp.core.CPOutcome;
import oscar.cp.core.CPPropagStrength;
import oscar.cp.core.CPVarInt;
import oscar.cp.core.Constraint;

/**
 * @author Pierre Schaus pschaus@gmail.com
 */
public class AllDiffFWC extends Constraint {

	private CPVarInt[] x;
	
	public AllDiffFWC(CPVarInt[] x) {
		super(x[0].s(),"AllDifferent forward checking");
		this.x = x;
	}

	@Override
	public CPOutcome setup(CPPropagStrength l) {	
		
		for (int i = 0; i < x.length; i++) {
			if(x[i].isBound()) {
				if(valBindIdx(x[i],i) == CPOutcome.Failure)
					return CPOutcome.Failure;
			}
		}
		for (int i = 0; i < x.length; i++) {
			if(!x[i].isBound()) {
				x[i].callValBindIdxWhenBind(this, i);
			}
		}
		return CPOutcome.Suspend;
	}
	
	public CPOutcome valBindIdx(CPVarInt x, int idx) {
		return filterBind(idx);
	}
	
	CPOutcome filterBind(int i) {
		assert(x[i].isBound());

		for (int j = 0; j < x.length; j++) {
			if (j != i) {
				if(x[j].removeValue(x[i].getValue()) == CPOutcome.Failure){
					return CPOutcome.Failure;
				}
			}
		}
		
		return CPOutcome.Suspend;
	}	

}

File: src/main/scala/oscar/cp/constraints/AllDiffFWC.scala
/*******************************************************************************
 * OscaR is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *   
 * OscaR is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License  for more details.
 *   
 * You should have received a copy of the GNU Lesser General Public License along with OscaR.
 * If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 ******************************************************************************/
package oscar.cp.constraints

import oscar.cp.core._
import oscar.reversible._
import oscar.cp.core.CPOutcome._

/**
 * Implementation of Sum Constraint:
 * @author Pierre Schaus pschaus@gmail.com
 */
class AllDiffFWC(val X: Array[CPVarInt]) extends Constraint(X(0).s, "AllDiffFWC") {

  val x = X.map(i => i)
  val nBounds = new ReversibleInt(s, 0)

  override def setup(l: CPPropagStrength): CPOutcome = {
    priorityL2 = CPStore.MAXPRIORL2 - 1
    X.foreach(_.callPropagateWhenBind(this, false))
    val oc = propagate()
    oc
  }

  private def setBound(i: Int) {
    val tmp = x(nBounds.value)
    x(nBounds.value) = x(i)
    x(i) = tmp
    nBounds.incr()
  }
  


  override def propagate(): CPOutcome = {

    var i = nBounds.value
    while (i < x.size) {
      if (x(i).isBound) {
        val v = x(i).value
        var j = nBounds.value
        while (j < x.size) {
          if (j != i && x(j).removeValue(v) == Failure) return Failure
          j += 1
        }
        setBound(i)
      }
      i += 1
    }
    Suspend
  }

}


